src/Product/ProductRepository.php

```diff
@@ -348,38 +351,91 @@ protected function get_presync_error_products_meta_query() {
-	 * - Set to not sync.
+	 * Find and return an array of WooCommerce product IDs that are marked as MC not_synced.
 	 * Excludes variations and variable products without variations.
 	 *
 	 * @param int $limit  Maximum number of results to retrieve or -1 for unlimited.
 	 * @param int $offset Amount to offset product results.
 	 *
 	 * @return int[] Array of WooCommerce product IDs
 	 */
-	public function find_not_synced_product_ids( int $limit = -1, int $offset = 0 ): array {
+	public function find_mc_not_synced_product_ids( int $limit = -1, int $offset = 0 ): array {
 		$types = ProductSyncer::get_supported_product_types();
 		$types = array_diff( $types, [ 'variation' ] );
 		$args  = [
 			'status'     => 'publish',
 			'type'       => $types,
 			'meta_query' => [
-				'relation' => 'OR',
-				$this->get_presync_error_products_meta_query(),
 				[
-					[
-						'key'     => ProductMetaHandler::KEY_VISIBILITY,
-						'compare' => '=',
-						'value'   => ChannelVisibility::DONT_SYNC_AND_SHOW,
-					],
+					'key'     => ProductMetaHandler::KEY_MC_STATUS,
+					'compare' => '=',
+					'value'   => MCStatus::NOT_SYNCED,
 				],
 			],
 		];
 
 		return $this->find_ids( $args, $limit, $offset );
 	}
 
+	/**
+	 * Find and return an array of WooCommerce product IDs that are NOT marked as MC not_synced.
+	 * Excludes variations and variable products without variations.
+	 *
+	 * @since x.x.x
+	 *
+	 * @param int $limit  Maximum number of results to retrieve or -1 for unlimited.
+	 * @param int $offset Amount to offset product results.
+	 *
+	 * @return int[] Array of WooCommerce product IDs
+	 */
+	public function find_mc_synced_product_ids( int $limit = -1, int $offset = 0 ): array {
+		$types = ProductSyncer::get_supported_product_types();
+		$types = array_diff( $types, [ 'variation' ] );
+		$args  = [
+			'status'     => 'publish',
+			'type'       => $types,
+			'meta_query' => [
+				[
+					'key'     => ProductMetaHandler::KEY_MC_STATUS,
+					'compare' => '!=',
+					'value'   => MCStatus::NOT_SYNCED,
+				],
+			],
+		];
+
+		return $this->find_ids( $args, $limit, $offset );
+	}
+
+	/**
+	 * Returns an array of Google Product IDs associated with all synced WooCommerce products.
+	 * Note: excludes variable parent products as only the child variation products are actually synced
+	 * to Merchant Center
+	 *
+	 * @since x.x.x
+	 *
+	 * @return array Google Product IDS
+	 */
+	public function find_all_synced_google_ids(): array {
+		// Don't include variable parent products as they aren't actually synced to Merchant Center.
+		$args['type']        = array_diff( ProductSyncer::get_supported_product_types(), [ 'variable' ] );
+		$synced_product_ids  = $this->find_synced_product_ids( $args );
+		$google_ids_meta_key = $this->prefix_meta_key( ProductMetaHandler::KEY_GOOGLE_IDS );
+		$synced_google_ids   = [];
+		foreach ( $synced_product_ids as $product_id ) {
+			$meta_google_ids = get_post_meta( $product_id, $google_ids_meta_key, true );
```

Comment:
This will still trigger all these posts to be loaded into cache/memory. It's at least only synced products and not all, we also load these posts later on anyways, but it kind of takes away from processing these returned ids in a batch.

Since we'll mostly have one ID per product now, can't we remove this function and pass the set of synced product IDs to process in batches, and retrieve the Google ID's within the batched loop?

Reply:
Interesting. I've made some changes to load the Google IDs within `MerchantStatuses` directly: 7bbab2ab92d74d886d8be2ed5dad700f61b6b1ee

In my tests, there wasn't a noticeably difference in execution time, but it does reduce `get_post_meta` calls (even if those products/posts are still eventually loaded during the process).